File: c++/stack/minStack.cpp
#include <iostream>
#include <stdio.h>
#include <limits.h>
#include <vector>
#include <string>
#include <map>
#include <set>

using namespace std;

class MinStack
{
public:
    vector<int> stack;
    vector<int> stackDesc;
    MinStack()
    {
    }

    void push(int val)
    {
        stack.push_back(val);
        if (stackDesc.size() == 0 || val <= stackDesc[stackDesc.size() - 1])
        {
            stackDesc.push_back(val);
        }
    }

    void pop()
    {
        if (stack.size() == 0)
        {
            return;
        }
        int lastElement = stack[stack.size() - 1];
        stack.erase(stack.begin() + (stack.size() - 1));
        if (lastElement == stackDesc[stackDesc.size() - 1])
        {
            stackDesc.erase(stackDesc.begin() + (stackDesc.size() - 1));
        }
        return;
    }

    int top()
    {
        if (stack.size() == 0)
        {
            return 0;
        }
        return stack[stack.size() - 1];
    }

    int getMin()
    {
        return stackDesc[stackDesc.size() - 1];
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

int main()
{
    MinStack *obj = new MinStack();
    obj->push(-2);
    obj->push(0);
    obj->push(-3);
    cout << obj->getMin() << endl;
    obj->pop();
    cout << obj->top() << endl;
    cout << obj->getMin() << endl;
    /*
    MinStack stack = MinStack();
    stack.push(-2);
    stack.push(0);
    stack.push(-3);
    cout << stack.getMin() << endl;
    stack.pop();
    stack.top();
    cout << stack.getMin() << endl;
    */
    return 0;
}

File: python/Stacks/minStack.py
from typing import List
import collections


class MinStack:

    def __init__(self):
        self.stack = []
        self.stackDesc = []
        return

    def push(self, val: int) -> None:
        if len(self.stackDesc) == 0 or self.stackDesc[len(self.stackDesc)-1] >= val:
            self.stackDesc.append(val)
        self.stack.append(val)
        return       

    def pop(self) -> None:
        if len(self.stack) == 0:
            return
        toret = self.stack[len(self.stack)-1]
        self.stack.pop(len(self.stack)-1)
        if toret == self.stackDesc[len(self.stackDesc)-1]:
            self.stackDesc.pop(len(self.stackDesc)-1)
        return

    def top(self) -> int:
        if len(self.stack) == 0:
            return
        toret = self.stack[len(self.stack)-1]
        return toret

    def getMin(self) -> int:
        if len(self.stackDesc) == 0: return
        return self.stackDesc[len(self.stackDesc)-1]
        


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

soln = MinStack()
print(soln.push(2))
print(soln.push(0))
print(soln.push(3))
print(soln.push(0))
print(soln.getMin())
print(soln.pop())
print(soln.getMin())
print(soln.pop())
print(soln.getMin())
print(soln.pop())
print(soln.getMin())


